src/pause/138CC0.c

```diff
@@ -103,19 +108,393 @@ s32 pause_badges_scroll_offset_x(s32 x) {
     return x;
 }
 
-INCLUDE_ASM(s32, "pause/138CC0", pause_badges_try_remove);
+s32 pause_badges_try_remove(s16 badgeID) {
+    s16 *currentSlot = gPlayerData.equippedBadges;
+    s16 *slotToRemove = currentSlot;
+    s32 result = 0;
+    s32 i;
+
+    if (badgeID == BADGE_INVALID) {
+        return 0;
+    }
+
+    for (i = 0; i < ARRAY_COUNT(gPlayerData.equippedBadges); i++, currentSlot++) {
+        if (badgeID == *currentSlot) {
+            slotToRemove = currentSlot;
+            result = 1;
+            break;
+        }
+    }
 
-INCLUDE_ASM(s32, "pause/138CC0", pause_badges_try_equip);
+    // Remove and shift array contents
+    for (; i < ARRAY_COUNT(gPlayerData.equippedBadges)-1; i++, currentSlot++) {
+        *currentSlot = *(currentSlot+1);
+    }
+    *currentSlot = 0;
 
-INCLUDE_ASM(s32, "pause/138CC0", pause_badges_draw_bp_orbs);
+    if (*slotToRemove == 0) {
+        result = 2;
+    }
+    return result;
+}
+
+typedef s32 BadgeEquipResult;
+enum BadgeEquipResults {
+    EQUIP_RESULT_ALREADY_EQUIPPED,
+    EQUIP_RESULT_NOT_ENOUGH_BP,
+    EQUIP_RESULT_TOO_MANY_BADGES,
+    EQUIP_RESULT_SUCCESS,
+};
+
+BadgeEquipResult pause_badges_try_equip(s16 badgeID) {
+    PlayerData* playerData = &gPlayerData;
+    s16 *badgeSlot = &playerData->equippedBadges[0];
+    s32 i;
+    s32 totalEquippedBP;
+
+    if (badgeID == BADGE_NONE_STANDIN) {
+        badgeID = 0;
+    }
+    if (badgeID == BADGE_INVALID) {
+        badgeID = 0;
+    }
+    if (badgeID == 0) {
+        return EQUIP_RESULT_NOT_ENOUGH_BP;
+    }
+
+    for (i = 0; i < ARRAY_COUNT(playerData->equippedBadges); i++, badgeSlot++) {
+        if (*badgeSlot == badgeID) {
+            return EQUIP_RESULT_ALREADY_EQUIPPED;
+        }
+    }
+
+    totalEquippedBP = pause_get_total_equipped_bp_cost();
+    if (badgeID != 0) {
+        u8 moveID = gItemTable[badgeID].moveID;
+        s32 requiredBP = totalEquippedBP + gMoveTable[moveID].costBP;
+
+        if (playerData->maxBP < requiredBP) {
+            return EQUIP_RESULT_NOT_ENOUGH_BP;
+        }
+    }
+
+    badgeSlot = &playerData->equippedBadges[0];
+    for (i = 0; i < ARRAY_COUNT(playerData->equippedBadges); i++, badgeSlot++) {
+        if (*badgeSlot == 0) {
+            *badgeSlot = badgeID;
+            break;
+        }
+    }
+    if (i == ARRAY_COUNT(playerData->equippedBadges)) {
+        return EQUIP_RESULT_TOO_MANY_BADGES;
+    } else {
+        return EQUIP_RESULT_SUCCESS;
+    }
+}
+
+void pause_badges_draw_bp_orbs(s32 orbState, s32 x, s32 y) {
+    s32 orbSize = 8;
+
+    switch (orbState) {
+        case 0:
+            pause_draw_rect(x * 4, y * 4, (x + orbSize) * 4, (y + orbSize) * 4, 0, 0, 0x100, 0x400, 0x400);
+            break;
+        case 1:
+            pause_draw_rect(x * 4, y * 4, (x + orbSize) * 4, (y + orbSize) * 4, 0, 0, 0, 0x400, 0x400);
+            break;
+        default:
+            pause_draw_rect(x * 4, y * 4, (x + orbSize) * 4, (y + orbSize) * 4, 0, 0, 0x200, 0x400, 0x400);
```

Comment:
Given the clipping args (the last 4 args of pause_draw_rect) are in pixels, I'd prefer them in decimal rather than hex